cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -184,23 +183,27 @@ private void restoreActiveRuns() {
                   }
                   try {
                     LOG.info("Found active run: {}", runRecordDetail.getProgramRunId());
+                    ProgramOptions programOptions =
+                        new SimpleProgramOptions(
+                            runRecordDetail.getProgramRunId().getParent(),
+                            new BasicArguments(runRecordDetail.getSystemArgs()),
+                            new BasicArguments(runRecordDetail.getUserArgs()));
+                    ProgramDescriptor programDescriptor = this.store.loadProgram(
+                        runRecordDetail.getProgramRunId().getParent());
                     if (runRecordDetail.getStatus() == ProgramRunStatus.PENDING) {
-                      runRecordMonitorService.addRequest(runRecordDetail.getProgramRunId());
+                      flowControlMonitorService.addRequest(runRecordDetail.getProgramRunId(),
```

Comment:
I don't think this is needed if the run is already in the database.

Reply:
This is required for emitting flow control metrics.

Replaces `addRequest`() with `emitFlowControlMetrics()` call.

Reply:
When active runs are restored (in [restoreActiveRuns](https://github.com/cdapio/cdap/blob/343101be0679793058bb7458430438db0521f40d/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java#L169)), the launching count metrics needs to be updated for runs in `Pending` state, otherwise metrics will show incorrect value.

This is only for consistency of launching count metrics.

Even if we don't update the launching count metrics, and a run request is made, it will update the DB and the launching count metric.

Reply:
Metrics we already emit in the end of startup once, no need to do it for every run

Reply:
Removed.

Yes, metrics are emitted [here](https://github.com/cdapio/cdap/blob/66da732da129e47bc72ff8f636239e064840a970/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java#L162 after all launches are restored.

Reply:
So, coming back, why do you need to call `flowControlService.addRequest`? Run is already in the database, no need to add it again